utils/docker_utils.go

```diff
@@ -0,0 +1,153 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"time"
+)
+
+type ContainerConfig struct {
+	Name   string
+	Image  string
+	Mounts []mount.Mount
+}
+
+func NewDockerClient(host string) *client.Client {
+	if err := os.Setenv("DOCKER_HOST", host); err != nil {
+		log.Print(err)
+	}
+	cli, _ := client.NewEnvClient()
+	return cli
+}
+
+func ListContainer(cli *client.Client) {
+	ctns, err := cli.ContainerList(context.Background(), types.ContainerListOptions{All: true})
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, container := range ctns {
+		fmt.Printf("%s %s\n", container.ID[:10], container.Image)
+	}
+}
+
+func CreateContainer(cli *client.Client, config ContainerConfig) (string, error) {
+	log.Printf("Creating Container [%s]", config.Name)
+	ctx := context.Background()
+	hostConfig := container.HostConfig{
+		Mounts: config.Mounts,
+	}
+
+	dconfig := container.Config{
+		Image: config.Image,
+		Tty:   true,
+	}
+
+	b, err := cli.ContainerCreate(ctx, &dconfig, &hostConfig, nil, config.Name)
+
+	return b.ID, err
+}
+
+func StartContainer(cli *client.Client, id string) error {
+	log.Printf("Starting Container [%s]", id)
+	return cli.ContainerStart(context.Background(), id, types.ContainerStartOptions{})
+}
+
+func StopContainer(cli *client.Client, id string) error {
+	log.Printf("Stopping Container [%s]", id)
+	var timeout = 5 * time.Second
+	return cli.ContainerStop(context.Background(), id, &timeout)
+}
+
+func RemoveContainer(cli *client.Client, id string) error {
+	log.Printf("Removing Container [%s]", id)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	return cli.ContainerRemove(ctx, id, types.ContainerRemoveOptions{})
+}
+
+func Write(cli *client.Client, id string, content []string, dest string) error {
+	for _, c := range content {
+		c = strings.ReplaceAll(c, "'", "'\"'\"'")
+		cmd := fmt.Sprintf(`echo -E '%s' >> %s`, c, dest)
+		log.Printf("Writing [%s] on [%s] from Container [%s]", c, dest, id)
+		Exec(cli, id, cmd)
+	}
+	return nil
+}
+
+func Copy(cli *client.Client, id, src, dest string) error {
+	log.Printf("Copy [%s] to [%s] from Container [%s]", src, dest, id)
+	dat, _ := ioutil.ReadFile(src)
+	content := string(dat)
+	content = strings.ReplaceAll(content, "'", "'\"'\"'")
+	cmd := fmt.Sprintf("echo -E '%s' >| %s", content, dest)
+	return Exec(cli, id, cmd)
+}
+
+func Exec(cli *client.Client, id, cmd string) error {
```

Comment:
I'm wondering about the trade-offs of executing commands one by one.

Reply:
The task commands you mean? If so, we do exec one by one, but inside the container (scenes of next prs).. The others commands we need to execute (to setup the environment, for example) are executed using this func, but I think it doesn't add any performance issue...